python/hail/linalg/blockmatrix.py

```diff
@@ -553,103 +604,207 @@ def __neg__(self):
         op = getattr(self._jbm, "unary_$minus")
         return BlockMatrix(op())
 
-    @typecheck_method(b=oneof(numeric, block_matrix_type))
+    def _check_common_dim(self, b, op):
+        if self.shape[0] != b.shape[0] and self.shape[1] != b.shape[1]:
+            raise ValueError(f'incompatible shapes for {op}: {self.shape} and {b.shape}')
+
+    def _promote(self, b, op):
+        a = self
+        form_a = Form.of(a.shape)
+        if isinstance(b, int) or isinstance(b, float):
+            b, form_b = float(b), Form.SCALAR
+        elif isinstance(b, np.ndarray):
+            b = _ndarray_as_2d_float64(b)
+            form_b = Form.of(b.shape)
+            if form_b == Form.SCALAR:
+                b = float(b[0][0])
+            elif form_a == Form.MATRIX and form_b in {Form.COLUMN, Form.ROW}:
+                a._check_common_dim(b, op)
+                b = _jarray_from_ndarray(b)
+            else:
+                if form_a in {Form.COLUMN, Form.ROW}:
+                    a._check_common_dim(b, op)
+                assert form_a <= form_b
+                return a._promote(BlockMatrix.from_numpy(b, self.block_size), op)
+        else:
+            assert isinstance(b, BlockMatrix)
+            form_b = Form.of(b.shape)
+            if form_b == Form.SCALAR:
+                b = b._as_scalar
+            elif form_a == Form.SCALAR:
+                a = a._as_scalar
+            elif form_a in {Form.COLUMN, Form.ROW} and form_b == Form.MATRIX:
+                a._check_common_dim(b, op)
+                a = a._jdata
+            elif form_a == Form.MATRIX and form_b in {Form.COLUMN, Form.ROW}:
+                a._check_common_dim(b, op)
+                b = b._jdata
+            else:
+                if a.shape != b.shape:
+                    raise ValueError(f'incompatible shapes for {op}: {a.shape} and {b.shape}')
+                if a.block_size != b.block_size:
+                    raise ValueError(f'block sizes disagree for {op}: {a.block_size} and {b.block_size}')
+
+        return a, b, form_a, form_b
+
+    @typecheck_method(b=oneof(numeric, np.ndarray, block_matrix_type))
     def __add__(self, b):
         """Addition: a + b.
 
         Parameters
         ----------
-        b: :class:`BlockMatrix` or :obj:`int` or :obj:`float`
+        b: :obj:`int` or :obj:`float` or :class:`numpy.ndarray` or :class:`BlockMatrix`
 
         Returns
         -------
         :class:`.BlockMatrix`
         """
-        if isinstance(b, int) or isinstance(b, float):
-            return BlockMatrix(self._jbm.scalarAdd(float(b)))
+        new_a, new_b, form_a, form_b = self._promote(b, 'addition')
+
+        if isinstance(new_a, float):
+            return BlockMatrix(new_b._jbm.scalarAdd(new_a))
```

Comment:
Was a wrong?  Can new_a be a float and not a block matrix?

Reply:
Both new_a and new_b can be block matrix, java array, or float.